system-configuration/src/network_configuration.rs

```diff
@@ -0,0 +1,452 @@
+                }
+            }
+        }
+
+        return None;
+    }
+
+    /// Returns primary network service
+    pub fn service(&self) -> Option<SCNetworkService> {
+        if let Some(service_id) = SCNetworkGlobal::query("ng_service", "PrimaryService") {
+            for _service in SCNetworkService::list() {
+                if _service.id() == service_id {
+                    return Some(_service);
+                }
+            }
+        }
+
+        return None;
+    }
+
+    /// Returns primary network interface
+    pub fn interface(&self) -> Option<SCNetworkInterface> {
+        if let Some(ifname) = SCNetworkGlobal::query("ng_interface", "PrimaryInterface") {
+            for iface in SCNetworkInterface::list() {
+                if let Some(bsd_name) = iface.bsd_name() {
+                    if bsd_name == ifname {
+                        return Some(iface);
+                    }
+                }
+            }
+        }
+
+        return None;
+    }
+
+    /// Returns default route on primary network service.
+    pub fn router(&self) -> Option<IpAddr> {
+        if let Some(router_str) = SCNetworkGlobal::query("ng_interface_router", "Router") {
+            if let Ok(router_ip) = router_str.parse::<IpAddr>() {
+                return Some(router_ip);
+            }
+        }
+
+        return None;
+    }
+
+    // pub fn netinfo(&self);
+    // pub fn proxies(&self) ;
+}
+
+/// Network service object.
+pub struct SCNetworkService(pub SCNetworkServiceRef);
+
+impl SCNetworkService {
+    /// Returns all available network services for the specified preferences.
+    pub fn list() -> Vec<SCNetworkService> {
+        let prefs = unsafe {
+            SCPreferencesCreate(
+                kCFAllocatorDefault,
+                CFString::from_static_string("ns_list").as_concrete_TypeRef(),
+                ptr::null(),
+            )
+        };
+
+        let array: CFArray<SCNetworkServiceRef> =
+            unsafe { CFArray::wrap_under_get_rule(SCNetworkServiceCopyAll(prefs)) };
+
+        array
+            .get_all_values()
+            .iter()
+            .map(|service_ptr| SCNetworkService(*service_ptr as _))
+            .collect::<Vec<SCNetworkService>>()
+    }
+
+    /// Returns the user-specified ordering of network services within the specified set.
+    pub fn list_order() -> Vec<SCNetworkService> {
+        let prefs = unsafe {
+            SCPreferencesCreate(
+                kCFAllocatorDefault,
+                CFString::from_static_string("ns_list_order").as_concrete_TypeRef(),
+                ptr::null(),
+            )
+        };
+
+        let netset = unsafe { SCNetworkSetCopyCurrent(prefs) };
+
+        let array: CFArray<SCNetworkServiceRef> =
+            unsafe { CFArray::wrap_under_get_rule(SCNetworkSetGetServiceOrder(netset)) };
+
+        let mut services = Vec::new();
+
+        for id in array.get_all_values().iter() {
+            let id_ptr: CFStringRef = *id as _;
```

Comment:
Here you treat `id` as a `CFStringRef`, but you have defined `array: CFArray<SCNetworkServiceRef>`, so at least one of them must be wrong :P. According to the documentation for `SCNetworkSetGetServiceOrder` it returns service identifiers, so I guess you want `CFArray<CFString>`?

If you just get the iterator for an array directly, without calling `get_all_values` then you get the correct type out of the iterator directly instead of just `*const c_void` which you then have to cast.

Reply:
Yeah, your are right :)